src/kOS/Sound/VoiceValue.cs

```diff
@@ -0,0 +1,149 @@
+            AddSuffix("VOLUME", new SetSuffix<ScalarValue>(() => voice.Volume, value => voice.Volume = value));
+            AddSuffix("WAVE", new SetSuffix<StringValue>(() => maker.GetWaveName(voiceNum), value => maker.SetWave(voiceNum, value.ToString())));
+            AddSuffix("PLAY", new OneArgsSuffix<Structure>(Play));
+            AddSuffix("LOOP", new SetSuffix<BooleanValue>(() => loop, value => loop = value));
+            AddSuffix("ISPLAYING", new SetSuffix<BooleanValue>(() => IsPlaying, value => IsPlaying = value));
+            AddSuffix("TEMPO", new SetSuffix<ScalarValue>(() => tempo, value => tempo = (float)value.GetDoubleValue()));
+        }
+
+        public void Play(Structure notes)
+        {
+            if (notes is NoteValue)
+            {
+                song = new ListValue();
+                song.Add(notes);
+            }
+            else if (notes is ListValue)
+            {
+                song = notes as ListValue;
+            }
+            else
+                throw new KOSInvalidArgumentException("Play", "note", "Requires either a NOTE() or a LIST() of NOTE()'s as its parameter.");
+
+            noteNum = -1;
+            noteEndTimeStamp = -1f;
+            IsPlaying = true;
+
+        }
+        
+        public void KOSUpdate(double deltaTime)
+        {
+            if (! IsPlaying)
+                return;
+
+            float now = Time.fixedTime;
+            
+            // If still playing prev note, do nothing except maybe change
+            // its frequency if it's a slidenote:
+            if (now < noteEndTimeStamp)
+            {
+                NoteValue note = song[noteNum] as NoteValue;
+                if (noteFreqTotalChange != 0.0)
+                {
+                    float durationPortion = (now - noteStartTimeStamp) / (noteEndTimeStamp - noteStartTimeStamp);
+                    float newFreq = note.Frequency + durationPortion*noteFreqTotalChange;
+                    voice.ChangeFrequency(newFreq);
+                }
+                return;
+            }
+            
+            // Increment to next note and start playing it:
+            ++noteNum;
+            if (noteNum > song.Count())
+            {
+                if (loop)
+                {
+                    noteNum = -1;
+                    noteEndTimeStamp = -1f;
+                }
+                else
+                    IsPlaying = false;
+            }
+            else
+            {
+                curNote = song[noteNum] as NoteValue;
+                if (curNote != null)
+                {
+                    noteStartTimeStamp = now;
+                    noteEndTimeStamp = now + tempo*curNote.Duration;
+                    noteFreqTotalChange = curNote.EndFrequency - curNote.Frequency;
+                    voice.BeginProceduralSound(curNote.Frequency, tempo*curNote.KeyDownLength, curNote.Volume);
+                }
```

Comment:
Yeah, I test using the debug instance of KSP, which pops up a red error message any time that there's an error, it really helps me find fringe cases like this.  In fact I have some other error messages that have popped up that I want to fix eventually too.